internal/auth/oidc/auth_method.go

```diff
@@ -0,0 +1,193 @@
+package oidc
+
+import (
+	"context"
+	"crypto/ed25519"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/base64"
+	"fmt"
+	"net/url"
+
+	"github.com/hashicorp/boundary/internal/auth/oidc/store"
+	"github.com/hashicorp/boundary/internal/errors"
+	"github.com/hashicorp/boundary/internal/kms"
+	"github.com/hashicorp/boundary/internal/oplog"
+	wrapping "github.com/hashicorp/go-kms-wrapping"
+	"github.com/hashicorp/go-kms-wrapping/structwrapping"
+	"google.golang.org/protobuf/proto"
+)
+
+// DefaultAuthMethodTableName defines the default table name for an AuthMethod
+const DefaultAuthMethodTableName = "auth_oidc_method"
+
+// AuthMethod contains an OIDC auth method configuration. It is owned
+// by a scope.  AuthMethods can have may assigned Accounts, AudClaims,
+// CallbackUrls, Certificates, SigningAlgs.  AuthMethods also have one State at
+// any given time which determines it's behavior for many its operations.
+type AuthMethod struct {
+	*store.AuthMethod
+	tableName string
+}
+
+// NewAuthMethod creates a new in memory AuthMethod assigned to scopeId.
+// WithMaxAge, WithName and WithDescription are the only valid options. All
+// other options are ignored.
+//
+// State equals the state of the OIDC auth method.  State is not a supported
+// parameter when creating new AuthMethod's since it must be Inactive for all
+// new AuthMethods.
+//
+// DiscoveryUrl equals a URL where the OIDC provider's configuration can be
+// retrieved. The OIDC Discovery Specification requires the URL end in
+// /.well-known/openid-configuration. However, Boundary will strip the URI and
+// only store the scheme and domain.
+//
+// ClientId equals an OAuth 2.0 Client Identifier valid at the Authorization
+// Server.
+//
+// ClientSecret equals the client's secret which will be encrypted when stored
+// in the database and an hmac representation will also be stored when every the
+// secret changes.  The secret is not returned via the API, the hmac is returned
+// so callers can determine if it's been updated.
+//
+// MaxAge equals the Maximum Authentication Age. Specifies the allowable elapsed
```

Comment:
What does it mean for the auth method's MaxAge to "force the IpD to re-authenticate the End-User"?  Does that mean the auth token is never generated or is generated but already expired?  Maybe this would be useful to test out the auth flow... but is there another time when that would be wanted?

Reply:
In short, the IdP must re-authenticate the user before issuing an auth token.  The token issued with have a "normal" TTL for the RP (defined by the IdP).  The issued auth token will have an `auth_time` claim, so the RP can verify the last time the end-user was authenticated. 

OIDC spec definition: _OPTIONAL: Maximum Authentication Age. Specifies the allowable elapsed time in seconds since the last time the end-user was actively authenticated by the OP. If the elapsed time is greater than this value, the OP must attempt to actively re-authenticate the end-user. (The max_age request parameter corresponds to the OpenID 2.0 PAPE max_auth_age request parameter.) When max_age is used, the ID token returned must include an auth_time claim value._

Reply:
By supporting `max_age`, we allow admins to support two things.

1. To enforce a minimum session freshness: If an app has a requirement that users must re-authenticate once per day, this can be enforced in the context of a much longer SSO session by providing max_age with a value. These are defined in seconds.
2. To force an immediate re-authentication: If an app requires that a user re-authenticate prior to access.